File: lab3.1.cpp
#include <iostream>
#include <cmath>   
#include <iomanip> 

using namespace std;

const double e = 1e-9;//эпсилон чтоб задать допустимую погрешность
//и сравнить числа с плавающей точкой
double f(double x_val, double gr_x, double peak_y) {
    double abs_x = abs(x_val);

    

    //граничные точки +/- 2
    if (abs(abs_x - gr_x) < e) {
        return 0.0;
    }
    //значения за пределами граничащих точек
    if (abs_x > gr_x) {
        return (abs_x - gr_x) / 3.0;
    }
 
    //основная функция ln(1/tg(x/2))
    double tg_arg = x_val / 2.0;
    double tg_val = tan(tg_arg);
    
    return log(abs(1.0 / tg_val));
}

int main() {
    double x_st = -5.0;
    double x_end = 5.0;
    double step = 0.5;

    double gr_x = 2.0; //функция меняет поведение   
    double peak_y = 3.0; //максимальный y в x=0
   
    cout << setw(10) << "x" << setw(10) << "y" << endl;
    
    //добавляем e/2.0 чтобы защитить от ошибок округления
    for (double x = x_st; x <= x_end + e / 2.0;x += step) {
        
        double y = f(x, gr_x, peak_y); 
        if(x ==0.0){
            cout << x << setw(10) << fixed << setprecision(4) << "inf" << endl;
            continue;
        }
        cout << setw(10) << fixed << setprecision(4) << x;

        cout << setw(10) << fixed << setprecision(4) << y << endl;
    }
    return 0; 
}

File: lab3.1.go
package main

import (
	"fmt"
	"math"
)

const (
	epsilon   = 1e-9  // допустимая погрешность
	xStart    = -5.0  // начальное значение x
	xEnd      = 5.0   // конечное значение x
	step      = 0.5   // шаг изменения x
	boundaryX = 2.0   // граница, где функция меняется
	peakY     = 3.0   // значение y при x = 0
)

// Функция, рассчитывающая значение y в зависимости от x
func computeY(x float64) float64 {
	absX := math.Abs(x)

	// x ≈ 0
	if absX < epsilon {
		return peakY
	}

	// |x| ≈ границе (например, 2.0)
	if math.Abs(absX-boundaryX) < epsilon {
		return 0.0
	}

	// |x| > границы: линейная часть
	if absX > boundaryX {
		return (absX - boundaryX) / 3.0
	}

	// Основная функция: y = ln(1 / |tg(x / 2)|)
	tgArg := x / 2.0
	tgVal := math.Tan(tgArg)

	return math.Log(math.Abs(1.0 / tgVal))
}

func main() {
	fmt.Printf("%10s%10s\n", "x", "y")

	for x := xStart; x <= xEnd+epsilon/2.0; x += step {
		y := computeY(x)
		fmt.Printf("%10.4f%10.4f\n", x, y)
	}
}
